### DISCUSSION TITLE
beginner question: z.infer()

### DISCUSSION THREAD

**Discussion by foreversun69** [colinhacks/zod#3165] [Q&A] [answered]
Hello,

i have a beginner question:
```js
const cb = () => z
    .function()
    .args(z.number())
    .returns(z.string());

const myCalculator1 = z
    .function()
    .args(z.number(), z.number(), cb()
    )
    .returns(z.string())
    .implement((num1, num2, cb) => {
        let sum = num1 + num2;
        let value = cb(sum);
        return value;
    });

type myFunction = z.infer<typeof myCalculator1>;
```

![image](https://github.com/colinhacks/zod/assets/44213573/57b19b56-abb7-401c-ab9c-c6b7cb0d3da8)
- what is here the problem?

best wishes.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const cb = () => z.function()
    .args( z.number() )
    .returns( z.string() )

const myCalculator1 = z.function()
    .args( z.number(), z.number(), cb() )
    .returns( z.string() )
    .implement( ( a, b, cb ) => cb( a + b ) )

type myFunction = ReturnType<typeof myCalculator1>
// type myFunction = string

console.log( myCalculator1( 1, 2, x => `${ x }` ) ) // 3
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by Ryan-Millard** [comment]
The issue here is with the `cb` function. You’re invoking it immediately when defining the arguments for myCalculator1. This means `cb` is not a function, but the result of the function, which is a Zod schema.

Here’s the corrected code:

const cb = z.function()
                    .args( z.number() ).returns( z.string() );

const myCalculator1 = z.function()
                                       .args( z.number(), z.number(), cb  )
                                       .returns( z.string() )
                                       .implement( (num1, num2, cb) => {
                                                             let sum = num1 + num2;
                                                             let value = cb(sum);
                                                             return value;
                                       });

type myFunction = z.infer<typeof myCalculator1>;

In this corrected version, `cb` is passed as a function schema to myCalculator1, not the result of the function. This allows myCalculator1 to correctly infer that its third argument should be a function that takes a number and returns a string.

**Reply by foreversun69** [reply]
Damn, thank you :-)
